### PR TITLE
refactor envelope to use cryptobytes

### PR DESCRIPTION
**Pull request by mikedanese**
A little less hairy.

```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go [hunk 1]
```diff
@@ -80,12 +80,11 @@ func (t *envelopeTransformer) TransformFromStorage(data []byte, context value.Co
...
 	// length cannot fit in 8 bits (1 byte). Thus, we use 16 bits (2 bytes) to store the length.
-	keyLen := int(binary.BigEndian.Uint16(data[:2]))
-	if keyLen+2 > len(data) {
-		return nil, false, fmt.Errorf("invalid data encountered by envelope transformer, length longer than available bytes: %q", data)
-	}
-	encKey := data[2 : keyLen+2]
-	encData := data[2+keyLen:]
+	var encKey cryptobyte.String
+	s := cryptobyte.String(data)
+	s.ReadUint16LengthPrefixed(&encKey)
```
Comment (awly) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope.go]: Check returned bool

**Review by awly — COMMENTED** [review]
Woah, that's a handy package

**Comment by liggitt** [comment]
that is a nice package. it looks like the unit test test round tripping using the TransformTo/TransformFrom, which is good to make sure they work with each other, but wouldn't catch if this had changed the representation in an incompatible way, right?

can you add a unit test that checks the output of TransformTo against specific bytes and make sure it passes against master as well?

**Comment by mikedanese** [comment]
Added a gross test to check back compat.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope_test.go [hunk 2]
```diff
@@ -201,3 +202,70 @@ func benchmarkRead(b *testing.B, transformer value.Transformer, valueLength int)
 	}
 	b.StopTimer()
 }
+
+// remove after 1.13
```
Comment (lavalamp) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/envelope_test.go]: This test is probably useful indefinitely to make sure the current implementation matches a known storage format, no?

**Pull request merged**